### PR TITLE
Allow customizing Vite plugins per router

### PR DESCRIPTION
**Pull request by XiNiHa**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
Users only can apply the same kind of Vite plugins to all Vinxi routers

## What is the new behavior?
Users can customize which Vite plugins to apply per Vinxi router

### CODE REVIEW

**Comment by nksaraf** [comment]
I like this and I think level of deetail about the infra is good to expose and will make people understand the system better. Because I don't think they can work well in it without understanding it. They need to know  the layers between vite and them. That's the public API that they are building against.

File: packages/start/config/index.d.ts [hunk 1]
```diff
@@ -9,7 +9,14 @@ type SolidStartInlineConfig = Omit<InlineConfig, "router"> & {
...
         islands?: boolean
+    },
+    plugins?: Plugins | {
+        client?: Plugins,
+        server?: Plugins,
+        serverFunctions?: Plugins
     }
 }
 
+type Plugins = InlineConfig["plugins"] | (() => Promise<InlineConfig["plugins"]> | InlineConfig["plugins"])
```
Comment (brenelz) [inline packages/start/config/index.d.ts:20]: you have `InlineConfig["plugins"]` twice?

Comment (XiNiHa) [reply packages/start/config/index.d.ts:20]: That's interpreted as `T | (() => (Promise<T> | T))`

File: packages/start/config/index.js [hunk 2]
```diff
@@ -57,6 +57,19 @@ export function defineConfig(baseConfig = {}) {
     entryExtension = ".jsx";
   }
 
+  const { clientPlugins, serverPlugins, serverFunctionsPlugins } =
+    typeof plugins === 'object' && !Array.isArray(plugins)
```
Comment (nksaraf) [inline packages/start/config/index.js:61]: should check if plugins is a function, if it is treat as the same as if it was an array

Comment (XiNiHa) [reply packages/start/config/index.js:61]: `typeof plugins` will return `"function"` if it is.

**Comment by ryansolid** [comment]
Hmm.. this does ask a question though. I'm starting to wonder if trying to wrap this into a Vite config was too ambitious. There are certain fields we can't support and things we need to expose from underlying layer.

**Comment by ryansolid** [comment]
Thanks for the PR but there is a new way to do this in 0.6.0

you can return a different vite config per router .. it is passed in as an argument to vite option.

**Pull request closed without merging**